pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -0,0 +1,1150 @@
+			glog.V(2).Infof("preStop hook for container %q did not complete in %d seconds", containerID, gracePeriod)
+		case <-done:
+			glog.V(3).Infof("preStop hook for container %q completed", containerID)
+		}
+		gracePeriod -= int64(unversioned.Now().Sub(start.Time).Seconds())
+	}
+
+	if gracePeriodOverride == nil {
+		// always give containers a minimal shutdown window to avoid unnecessary SIGKILLs
+		if gracePeriod < minimumGracePeriodInSeconds {
+			gracePeriod = minimumGracePeriodInSeconds
+		}
+	} else {
+		gracePeriod = *gracePeriodOverride
+		glog.V(2).Infof("Killing container %q, but using %d second grace period override", containerID, gracePeriod)
+	}
+
+	err := m.runtimeService.StopContainer(containerID.ID, gracePeriod)
+	if err != nil {
+		glog.V(2).Infof("Container %q termination failed after %s: %v", containerID.String(), unversioned.Now().Sub(start.Time), err)
+	} else {
+		glog.V(2).Infof("Container %q exited after %s", containerID.String(), unversioned.Now().Sub(start.Time))
+	}
+
+	ref, ok := m.containerRefManager.GetRef(containerID)
+	if !ok {
+		glog.Warningf("No ref for container %s", containerID.String())
+	} else {
+		message := fmt.Sprintf("Killing container with id %s", containerID.String())
+		if reason != "" {
+			message = fmt.Sprint(message, ":", reason)
+		}
+		m.recorder.Event(ref, api.EventTypeNormal, events.KillingContainer, message)
+		m.containerRefManager.ClearRef(containerID)
+	}
+
+	return err
+}
+
+func (m *kubeGenericRuntimeManager) killPodWithSyncResult(pod *api.Pod, runningPod kubecontainer.Pod, gracePeriodOverride *int64) (result kubecontainer.PodSyncResult) {
+	containerResults := make(chan *kubecontainer.SyncResult, len(runningPod.Containers))
+	wg := sync.WaitGroup{}
+
+	wg.Add(len(runningPod.Containers))
+	for _, container := range runningPod.Containers {
+		go func(container *kubecontainer.Container) {
+			defer utilruntime.HandleCrash()
+			defer wg.Done()
+
+			var containerSpec *api.Container
+			if pod != nil {
+				for i, c := range pod.Spec.Containers {
+					if container.Name == c.Name {
+						containerSpec = &pod.Spec.Containers[i]
+						break
+					}
+				}
+			}
+
+			killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, container.Name)
+			if err := m.killContainer(pod, container.ID, containerSpec, "Need to kill Pod", gracePeriodOverride); err != nil {
+				killContainerResult.Fail(kubecontainer.ErrKillContainer, err.Error())
+			}
+			containerResults <- killContainerResult
+		}(container)
+	}
+
+	wg.Wait()
+	close(containerResults)
+	for containerResult := range containerResults {
+		result.AddSyncResult(containerResult)
+	}
+
+	// Teardown network plugin
+	podName := runningPod.Name
+	podNamespace := runningPod.Namespace
+	if len(podName) == 0 && pod != nil {
+		podName = pod.Name
+		podNamespace = pod.Namespace
+	}
+
+	podFullName := kubecontainer.BuildPodFullName(podName, podNamespace)
+	found, podSandboxID, err := m.getPodSandboxID(podFullName, nil)
```

Comment:
As @Random-Liu mentioned above, `podFullName` is not sufficient to identify a sandbox. We may have already stopped the sandbox and started a new one. Will it make more sense to just use pod UID to filter sandboxes? That's what we are doing with docker today.